### PR TITLE
[DEVOPS-10] Windows build automation

### PR DESCRIPTION
**Pull request by deepfire**
- move most contents of the `appveyor.yml` build script into `scripts/windows-build-fresh-daedalus.bat`, adding some sanity checks on the way
- scripts/windows-build-fresh-daedalus.bat, for a shortcuttable easy builds on Windows

### CODE REVIEW

File: scripts/build-installer-win64.bat [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+@echo Obtaining Cardano from branch %CARDANO_BRANCH%
+rmdir /s/q node_modules\daedalus-client-api 2>nul
+mkdir      node_modules\daedalus-client-api
+
+pushd node_modules\daedalus-client-api
+    powershell -Command "try { $wc = New-Object net.webclient; $wc.Downloadfile('https://ci.appveyor.com/api/projects/jagajaga/cardano-sl/artifacts/CardanoSL.zip?branch=%CARDANO_BRANCH%', 'CardanoSL.zip'); } catch { exit 1; }"
+    @if %errorlevel% neq 0 (@echo FAILED: powershell -Command "try { $wc = New-Object net.webclient; $wc.Downloadfile('https://ci.appveyor.com/api/projects/jagajaga/cardano-sl/artifacts/CardanoSL.zip?branch=%CARDANO_BRANCH%', 'CardanoSL.zip'); } catch { exit 1; }"
+	popd & exit /b 1)
+    @for /F "usebackq" %%A in ('CardanoSL.zip') do set size=%%~zA
+    if %size% lss %MIN_CARDANO_BYTES% (@echo FAILED: CardanoSL.zip is too small: threshold=%MIN_CARDANO_BYTES%, actual=%size% bytes
```
Comment (jmitchell) [inline scripts/build-installer-win64.bat]: Very nice improvement 👍

File: scripts/build-installer-win64.bat [hunk 2]
```diff
@@ -0,0 +1,102 @@
+rem DEPENDENCIES:
+rem   1. Node.js ('npm' binary in PATH)
+rem   2. 7zip    ('7z'  binary in PATH)
+
+set DEFAULT_CARDANO_BRANCH=cardano-sl-0.4
```
Comment (jmitchell) [inline scripts/build-installer-win64.bat]: Usually we want the Cardano dependency branch to match the current Daedalus branch. This is good, but you might consider setting the default branch that way.

Currently the AppVeyor script doesn't pass the 2nd argument, so it relies on this default.

File: scripts/build-installer-win64.bat [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+set MIN_CARDANO_BYTES=50000000
+set LIBRESSL_VERSION=2.5.3
+
+set DAEDALUS_VERSION=%1
```
Comment (jmitchell) [inline scripts/build-installer-win64.bat]: Perhaps default to `0.0.0` or `dev`.

File: scripts/build-installer-win64.bat [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    del CardanoSL.zip
+popd
+
+move   node_modules\daedalus-client-api\log-config-prod.yaml installers\log-config-prod.yaml
+move   node_modules\daedalus-client-api\cardano-node.exe     installers\
+move   node_modules\daedalus-client-api\cardano-launcher.exe installers\
+del /f node_modules\daedalus-client-api\*.exe
+
+rem @echo Obtaining LibreSSL %LIBRESSL_VERSION%
+rem https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-%LIBRESSL_VERSION%-windows.zip
```
Comment (jmitchell) [inline scripts/build-installer-win64.bat]: Is the plan to eventually install libressl automatically, perhaps conditional upon a var? If so, consider adding a comment indicating the plan.

The thing about commented code w/o an explanation is people don't know whether it indicates past failure or unfulfilled future plans.

File: scripts/build-installer-win64.bat [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    @if %errorlevel% neq 0 (
+        @echo .
+        @echo .
+        @echo FAILED: stack --no-terminal build -j 2 --exec make-installer
+        @echo .
+        @timeout 7
+        @echo Retrying -- and also waiting for GHC 8.2.1 [see https://github.com/commercialhaskell/stack/issues/2617]
+        @echo .
+        @echo .
+	goto build
```
Comment (jmitchell) [inline scripts/build-installer-win64.bat]: This is an infinite loop, right? We need to bail after several failed tries, otherwise CI may spin its wheels on failing code.

File: scripts/windows-build-fresh-daedalus.bat [hunk 6]
```diff
@@ -0,0 +1,29 @@
+rem DEPENDENCIES:
+rem   1. Node.js ('npm' binary in PATH)
+rem   2. 7zip    ('7z'  binary in PATH)
+rem   3. Git     ('git' binary in PATH)
+
+@set DEFAULT_BRANCH=cardano-sl-0.4
```
Comment (jmitchell) [inline scripts/windows-build-fresh-daedalus.bat]: Again, we may want to infer this from the current Daedalus branch.

File: scripts/build-installer-win64.bat (hunk 3 above)
Comment (jmitchell) [inline scripts/build-installer-win64.bat]: Hmm, maybe not, actually. I see your convenient local build script `windows-build-fresh-daedalus.bat` handles this.

File: scripts/build-installer-win64.bat (hunk 4 above)
Comment (deepfire) [inline scripts/build-installer-win64.bat]: It's dealt with in the next PR : -)

File: scripts/build-installer-win64.bat (hunk 5 above)
Comment (deepfire) [inline scripts/build-installer-win64.bat]: Oh, right -- the retry logic in  Appveyor has an iteration limit, indeed.

File: scripts/build-installer-win64.bat (hunk 2 above)
Comment (deepfire) [inline scripts/build-installer-win64.bat]: @jmitchell, ok, so:
1. making `appveyor.yml` and `windows-build-fresh-daedalus.bat` specify the version
2. making the second argument (cardano version) mandatory in `scripts/build-installer-win64.bat`

File: scripts/windows-build-fresh-daedalus.bat (hunk 6 above)
Comment (deepfire) [inline scripts/windows-build-fresh-daedalus.bat]: This is different -- this script is supposed to live outside the repository -- since it's responsible for a convenient developer workflow -- shortcutting it allows for a single-click end-to-end fresh installer build, basically.

And since it lives outside the repository.. well, any notion of "current" is removed not by one, but by two steps:
1. finding out the repository location
2. inferring the Cardano branch -- which is a problem of its own

In short, I don't see how we can do much here..

**Comment by deepfire** [comment]
@jmitchell, I hope I have addressed your concerns..

**Comment by domenkozar** [comment]
We need to give you write access, so we can test that produced artifact indeed is done.

**Comment by deepfire** [comment]
Resubmitting this as as new PR https://github.com/input-output-hk/daedalus/pull/284, due to the CI integration issues.
